rustls/src/compress.rs

```diff
@@ -223,6 +229,144 @@ mod feat_brotli {
+        }
+
+        // cache probe:
+        let encoding = original.get_encoding();
+        let algorithm = compressor.algorithm();
+
+        let mut cache = self
+            .inner
+            .lock()
+            .map_err(|_| CompressionFailed)?;
+        for (i, item) in cache.iter().enumerate() {
+            if item.algorithm == algorithm && item.original == encoding {
+                // this item is now MRU
+                let item = cache.remove(i);
+                cache.push(Arc::clone(&item));
+                return Ok(item);
+            }
+        }
+        drop(cache);
+
+        // do compression:
+        let uncompressed_len = encoding.len() as u32;
+        let compressed = compressor.compress(encoding.clone(), CompressionLevel::Amortized)?;
+        let new_entry = Arc::new(CompressionCacheEntry {
+            algorithm,
+            original: encoding,
+            compressed: CompressedCertificatePayload {
+                alg: algorithm,
+                uncompressed_len,
+                compressed: PayloadU24(Payload::new(compressed)),
+            },
+        });
+
+        // insert into cache
+        let mut cache = self
+            .inner
+            .lock()
+            .map_err(|_| CompressionFailed)?;
+        if cache.len() == self.size {
+            cache.remove(0);
+        }
+        cache.push(Arc::clone(&new_entry));
+        Ok(new_entry)
+    }
+
+    #[cfg(not(feature = "std"))]
+    /// Create a dummy compression cache.
+    ///
+    /// The cache is not supported for no_std builds.
+    pub fn new(_unused: usize) -> Self {
+        Self {}
+    }
+
+    #[cfg(not(feature = "std"))]
+    pub(crate) fn compression_for(
+        &self,
+        compressor: &dyn CertCompressor,
+        original: &CertificatePayloadTls13,
+    ) -> Result<Arc<CompressionCacheEntry>, CompressionFailed> {
+        Self::uncached_compression(compressor, original)
+    }
+
+    /// Compress `original` using `compressor`, but do not insert the result into the cache.
+    fn uncached_compression(
+        compressor: &dyn CertCompressor,
+        original: &CertificatePayloadTls13,
+    ) -> Result<Arc<CompressionCacheEntry>, CompressionFailed> {
+        let algorithm = compressor.algorithm();
+        let encoding = original.get_encoding();
+        let uncompressed_len = encoding.len() as u32;
+        let compressed = compressor.compress(encoding, CompressionLevel::Interactive)?;
+
+        // this `CompressionCacheEntry` in fact never makes it into the cache, so
+        // `original` is left empty
+        Ok(Arc::new(CompressionCacheEntry {
+            algorithm,
+            original: Vec::new(),
+            compressed: CompressedCertificatePayload {
+                alg: algorithm,
+                uncompressed_len,
+                compressed: PayloadU24(Payload::new(compressed)),
+            },
+        }))
+    }
+}
+
+#[cfg_attr(not(feature = "std"), allow(dead_code))]
```

Comment:
I'm surprised this dead code allow is needed for no-std, isn't this fn ref'd by `pub(crate) compression_for(...)` in the `no-std` case?

Reply:
The complaint this addresses is about `CompressionCacheEntry::algorithm` and `CompressionCacheEntry::original` being written but never read.

Reply:
Ahh! Ok that makes sense.

I found it's a little awkward to be using the `CompressionCacheEntry` for both the cached use case (where these fields are read, and the `original` is non-empty) and the non-cached case (where the fields are never read, and `original` is always empty) but I couldn't think of a better approach.

I think it's OK as-is. Thanks for explaining the `allow`.